quantfinlib/train_test_split/kfoldpe.py

```diff
@@ -0,0 +1,197 @@
+            groups = np.arange(X.shape[0])
+        indices = np.arange(X.shape[0])
+        assert len(groups) == len(indices), "Bummer! Groups and indices must have the same length."
+        test_indices_per_fold = np.array_split(indices, self.n_folds)
+        # fix the bounds of the folds so that there is no overlap between their groups
+        bnd_per_fold = np.array([[test_indices[0], test_indices[-1]] for test_indices in test_indices_per_fold])
+        bnd_per_fold = modify_fold_bnds(bnd_per_fold=bnd_per_fold, groups=groups)
+        test_indices_per_fold = [np.arange(bnd[0], bnd[1] + 1) for bnd in bnd_per_fold]
+        if self.look_forward:
+            test_indices_per_fold = test_indices_per_fold[1:]
+        for test_index in test_indices_per_fold:
+            train_index = get_train_index_per_split(
+                indices=indices, test_index=test_index, groups=groups, look_forward=self.look_forward
+            )
+            train_index = _purge(
+                train_index=train_index, test_index=test_index, indices=indices, groups=groups, n_purge=self.n_purge
+            )
+            train_index = _embargo(
+                train_index=train_index, test_index=test_index, indices=indices, groups=groups, n_embargo=self.n_embargo
+            )
+            yield train_index, test_index
+
+
+def modify_fold_bnds(bnd_per_fold: np.ndarray, groups: np.ndarray) -> np.ndarray:
+    """
+    Modify fold bounds.
+
+    Modifies the bounds of folds such that there is no overlap between their groups.
+
+    Parameters
+    ----------
+    bnd_per_fold : np.ndarray
+        Bound per fold.
+    groups : np.ndarray
+        Groups.
+    n_folds : int
+        Number of folds.
+
+    Returns
+    -------
+    np.ndarray
+        Modified bound per fold.
+    """
+    n_folds = bnd_per_fold.shape[0]
+    for fold in range(n_folds - 1):
+        last_group_curr_fold = groups[bnd_per_fold[fold, 1]]
+        first_group_next_fold = groups[bnd_per_fold[fold + 1, 0]]
+        if last_group_curr_fold == first_group_next_fold:
+            next_group = groups[groups > last_group_curr_fold][0]
+            bnd_per_fold[fold, 1] = np.where(groups == last_group_curr_fold)[0][-1]
+            bnd_per_fold[fold + 1, 0] = np.where(groups == next_group)[0][0]
+    return bnd_per_fold
+
+
+def get_train_index_per_split(
+    indices: np.ndarray,
+    test_index: np.ndarray,
+    groups: Union[np.ndarray, pd.TimedeltaIndex],
+    look_forward: bool = False,
+) -> np.ndarray:
+    """Find train index for a given split.
+
+    Parameters
+    ----------
+    indices : np.ndarray
+        All indices.
+    test_index : np.ndarray
+        Test indices.
+    groups : Union[np.ndarray, pd.TimedeltaIndex]
+        Groups.
+    look_forward : bool
+        Whether to test set is always ahead of the training set.
+
+    Returns
+    -------
+    np.ndarray
+        Train indices for a given split.
+    """
+    train_index = np.setdiff1d(indices, test_index)

```

Comment:
now train_index is computed twice in case look_forward is True

Reply:
fixe it.